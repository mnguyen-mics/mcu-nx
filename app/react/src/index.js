import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';

import configureStore from './store';
import { Navigator } from './containers/Navigator';

const store = configureStore();

const micsProvider = (
  <Provider store={store}>
    <Navigator />
  </Provider>
);

render(
  micsProvider,
  document.getElementById('mcs-react-app') // eslint-disable-line no-undef
);
