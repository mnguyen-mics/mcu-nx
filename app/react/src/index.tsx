/* eslint-disable */
import React from 'react';
import { render } from 'react-dom';

import App from './App';

declare global {
  interface Window {
    require: any;
    mainRev: any;
  }
}

render(<App />, document.getElementById('mcs-react-app'), () => {
  window.require([window.mainRev], () => {
    //
  });
});
