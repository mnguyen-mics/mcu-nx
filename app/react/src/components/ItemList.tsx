import * as React from 'react';
import { withRouter, RouteComponentProps } from 'react-router';
import { compose } from 'recompose';

import { EmptyTableView, TableViewFilters } from './TableView';
import { FormattedMessage } from 'react-intl';
import { McsIconType } from './McsIcon';
import {
  buildDefaultSearch,
  compareSearches,
  isSearchValid,
  parseSearch,
  updateSearch,
} from '../utils/LocationSearchHelper';
import { ViewComponentWithFiltersProps } from './TableView/TableViewFilters';

export interface Filters {
  currentPage?: number;
  pageSize?: number;
  status?: string[];
  keywords?: string;
}

interface RouterParams {
  organisationId: string;
}

export interface PageSetting {
  paramName: string;
  defaultValue: number;
  deserialize: (query: any) => number;
  serialize: (value: string | string[] | number) => string;
  isValid: (query: any) => boolean;
}

interface EmptyTableProps {
  iconType: McsIconType;
  intlMessage: FormattedMessage.Props;
}

export interface ItemListProps<T = any> extends ViewComponentWithFiltersProps<T> {
  fetchList: (organisationId: string, filters: Filters, isInitialFetch?: boolean) => void;
  dataSource: T[];
  total: number;
  pageSettings: PageSetting[];
  emptyTable: EmptyTableProps;
  additionnalComponent?: React.ReactNode;
}

type Props<T = any> = ItemListProps<T> & RouteComponentProps<RouterParams>

class ItemList<T> extends React.Component<Props<T>> {

  componentDidMount() {
    const {
      fetchList,
      history,
      location: { search, pathname },
      match: { params: { organisationId } },
      pageSettings,
    } = this.props;

    if (!isSearchValid(search, pageSettings)) {
      history.replace({
        pathname: pathname,
        search: buildDefaultSearch(search, pageSettings),
        state: { reloadDataSource: true },
      });
    } else {
      const filters = parseSearch(search, pageSettings);

      fetchList(organisationId, filters, true);
    }
  }

  componentWillReceiveProps(nextProps: Props<T>) {
    const {
      fetchList,
      history,
      location: {
        search,
      },
      match: {
        params: {
          organisationId,
        },
      },
    } = this.props;

    const {
      location: {
        pathname: nextPathname,
        search: nextSearch,
      },
      match: {
        params: {
          organisationId: nextOrganisationId,
        },
      },
      pageSettings,
    } = nextProps;

    if (!compareSearches(search, nextSearch) || organisationId !== nextOrganisationId) {
      if (!isSearchValid(nextSearch, pageSettings)) {
        history.replace({
          pathname: nextPathname,
          search: buildDefaultSearch(nextSearch, pageSettings),
          state: { reloadDataSource: organisationId !== nextOrganisationId },
        });
      } else {
        const filters = parseSearch(nextSearch, pageSettings);
        fetchList(nextOrganisationId, filters, false);
      }
    }
  }

  updateLocationSearch = (params: Filters) => {
    const {
      history,
      location: { search: currentSearch, pathname },
      pageSettings,
    } = this.props;

    const nextLocation = {
      pathname,
      search: updateSearch(currentSearch, params, pageSettings),
    };

    history.push(nextLocation);
  }

  render() {
    const {
      total,
      location: { search },
      pageSettings,
      emptyTable: {
        iconType,
        intlMessage,
      },
      additionnalComponent,
      ...rest
    } = this.props;

    if (!rest.dataSource.length && !rest.loading && !additionnalComponent) {
      return <EmptyTableView iconType={iconType} intlMessage={intlMessage} />;
    }

    const filter = parseSearch(search, pageSettings);
    const pagination = {
      current: filter.currentPage,
      pageSize: filter.pageSize,
      onChange: (page: number) =>
        this.updateLocationSearch({
          currentPage: page,
        }),
      onShowSizeChange: (current: number, size: number) =>
        this.updateLocationSearch({
          currentPage: 1,
          pageSize: size,
        }),
      total,
    };

    return (
      <div className="mcs-table-container">
        {additionnalComponent}
        {!rest.dataSource.length && !rest.loading ? (
          <EmptyTableView
            iconType={iconType}
            intlMessage={intlMessage}
            className="mcs-table-view-empty mcs-empty-card"
          />
        ) : (
          <TableViewFilters pagination={pagination} {...rest} />
        )}
      </div>
    );
  }
}

export default compose<Props, ItemListProps>(
  withRouter,
)(ItemList);
