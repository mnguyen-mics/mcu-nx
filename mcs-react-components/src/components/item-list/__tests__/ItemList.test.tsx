import 'jest';
import * as React from "react";
import { MemoryRouter } from 'react-router'
import { IntlProvider } from 'react-intl';
import * as TestRenderer from 'react-test-renderer';
import ItemList, { ItemListProps, Filters } from "../ItemList";

it("renders the item list", () => {
  const p = (organisationId: string, filters: Filters, isInitialFetch?: boolean) => { };
  const props: ItemListProps<any> = {
    selectionNotifyerMessages: {
      allRowsSelected: "string",
      unselectAll: "string",
      allPageRowsSelected: "string",
      selectAll: "string",
      selectedRows: "string",
    },
    fetchList: p,
    dataSource: [],
    total: 10,
    pageSettings: [],
    emptyTable: {
      iconType: 'email',
      message: "Empty Table"
    },
  };

  const component = TestRenderer.create(
    <IntlProvider locale="en">
      <MemoryRouter>
        <ItemList {...props} />
      </MemoryRouter>
    </IntlProvider>
  );
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot();
});
