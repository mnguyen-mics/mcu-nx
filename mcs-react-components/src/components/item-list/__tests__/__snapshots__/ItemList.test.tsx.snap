// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the item list 1`] = `
<div
  className="mcs-empty-table-view"
>
  <div
    className="ant-col ant-col-24 mcs-empty-table-view-content mcs-table-view-empty"
    style={Object {}}
  >
    <div
      className="mcs-empty-table-view-logo"
    >
      <span
        className="mcs-icon "
      >
        <i
          className="mcs-email"
          style={Object {}}
        />
      </span>
    </div>
    Empty Table
  </div>
</div>
`;
